### PR TITLE
Laboratory work #1, Maxim Krivenko - 23FPL2

### PR DESCRIPTION
**Pull request by makskrivenko**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -7,6 +7,18 @@
 
 
 def tokenize(text: str) -> list[str] | None:
+
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:11]: `if not` -> return `None` 
Return bad values as soon as possible.

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -7,6 +7,18 @@
...
+        text = text.lower()
+        tokens = []
+        for i in text:
+            if i.isalpha() and i != "º":
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:15]: Do we really need this special check?

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -23,6 +35,20 @@ def tokenize(text: str) -> list[str] | None:
 
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
+
+    if not isinstance(tokens, list):
+        return None
+    else:
+        for i in tokens:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:42]: try `any` or `all`

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -37,6 +63,16 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
 
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
+
+    if isinstance(language, str) and isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:67]: the same - return None as soon as possible

File: lab_1_classify_by_unigrams/start.py [hunk 5]
```diff
@@ -3,20 +3,26 @@
 """
 # pylint:disable=too-many-locals, unused-argument, unused-variable
 
+import main as func
+
 
 def main() -> None:
     """
-    Launches an implementation
+    Launches an i mplementation
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:11]: return

File: seminars/__init__.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+print("something")
```
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply seminars/__init__.py:1]: Hello, I do not want to see it the next time

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (makskrivenko) [reply lab_1_classify_by_unigrams/main.py:42]: Вместо всей конструкции?

File: lab_1_classify_by_unigrams/__init__.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+s = "name"
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/__init__.py:1]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -7,6 +7,19 @@
 
 
 def tokenize(text: str) -> list[str] | None:
+
+    if not isinstance(text, str):
+        return None
+
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:14]: `else` is unnecessary

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -37,6 +60,17 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
 
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
+
+    if not (isinstance(language, str) and isinstance(text, str)):
+        return None
+
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:67]: `unnecessary`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
every, EVERY code inside function must be AFTER docstring

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -34,6 +44,14 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not (isinstance(tokens, list) and all(isinstance(i, str) for i in tokens)):
+        return None
+
+    freq_dict = {}
+    for letter in tokens:
+        if letter not in freq_dict:
+            freq_dict[letter] = tokens.count(letter) / len(tokens)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:53]: You calculate the same value many times for similar symbols. Can you optimize it and calculate it only for unique

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -64,6 +92,17 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     """
+    if len(actual) != len(predicted):
+        return None
+
+    num_sum = 0
+
+    for actual_num, predicted_num in zip(actual, predicted):
+        num_sum += (actual_num - predicted_num) ** 2
+
+    mse = num_sum / len(actual)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:103]: do not create the variable - return `as is`

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -84,6 +123,30 @@ def compare_profiles(
...
+        if elements not in unknown_profile['freq']:
+            unknown_lst.append(float(0))
+        else:
+            unknown_lst.append(unknown_profile['freq'][elements])
+        if elements not in profile_to_compare['freq']:
+            compare_lst.append(float(0))
+        else:
+            compare_lst.append(profile_to_compare['freq'][elements])
+
+    comparison = calculate_mse(unknown_lst, compare_lst)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:147]: the same - do not create the variale

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -105,6 +168,30 @@ def detect_language(
...
+
+    if mse_1 is None or mse_2 is None:
+        return None
+
+    language = ''
+    equal_lst = [profile_1['name'], profile_2['name']]
+    equal_lst.sort()
+
+    if mse_1 > mse_2:
+        language = profile_2['name']
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:188]: return

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -105,6 +168,30 @@ def detect_language(
...
+    if mse_1 < mse_2:
+        language = profile_1['name']
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:190]: return

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -105,6 +168,30 @@ def detect_language(
...
+    if mse_1 == mse_2:
+        language = equal_lst[0]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:192]: return

File: seminars/practice_3_lists.py [hunk 16]
```diff
@@ -15,27 +15,24 @@
 
 # Create a list
 example = [1, 2, 3]
-print(example)
+
```
Comment (artyomtugaryov) [inline seminars/practice_3_lists.py:18]: revert

**Pull request closed without merging**